genienlp/data_utils/bootleg.py

```diff
@@ -141,6 +122,34 @@ def __init__(self, args):
             'False',
         ]
 
+    def create_wiki2normalized_type_mapping(self):
+        # keys are normalized types for each thingtalk property, values are a list of wiki types
+        self.almond_type_mapping = dict()
+    
+        # a list of tuples: each pair includes a wiki type and their normalized type
+        self.wiki2normalized_type = list()
+    
+        if self.args.almond_type_mapping_path:
+            # read mapping from user-provided file
+            with open(os.path.join(self.args.root, self.args.almond_type_mapping_path)) as fin:
+                self.almond_type_mapping = ujson.load(fin)
+            for normalized_type, titles in self.almond_type_mapping.items():
+                for title in titles:
+                    self.wiki2normalized_type.append((title, normalized_type))
```

Comment:
My suggestion was to take the parts common to the two branches of the if statement and make them into a method. Moving the whole thing into a method doesn't help (as much): you still have the same complexity in terms of nested control flow and redundant code.

Reply:
Only shared code between the two branches is this three line of code: https://github.com/stanford-oval/genienlp/blob/58aecfc46e70df5b2069d15292ccfd1ef9282c03/genienlp/data_utils/bootleg.py#L149-L151
Do you really want to refactor them into a method?

Reply:
Yes because those three lines of code have two loops in them, and they nested within other control flow. Have you heard of cyclomatic complexity? It's basically a measure of how nested your if statements & for loops go, and the higher the number the less readable code is.